### PR TITLE
Docker Swarm addon

### LINKED ISSUE #1406: Docker Swarm on top of Mantl
## Why
I propose that support for Docker Swarm should be added to Mantl.
By having Swarm on Mantl, all the (available) resources on a cluster are exposed via the Docker API directly. Developers are already used to this API / model of working.

Swarm is ideal to run one-time-jobs in production easily. Think about database migrations (which could be run detached), or running some interactive script to generate a report (in which the attached input/output is really usefull).

Finally, Swarm could allow for easy development of microservices on top of Mantl. A developer could write the code for microservices on his local MackBook Air, and use Docker Compose to start a set of microservices via Swarm on top of Mantl, which would use way more memory and CPU than his MacBook Air could provide.

## How
Swarm registers itself as a Mesos framework, uses it to allocate resources and uses the Docker Executor to start containers. So far, so good! However, it also requires access to the Docker API on each worker node to provide services like pulling/pushing of images and showing the logs. At the moment, only the `/var/run/sock` unix domain socket is enabled in Mantl. To expose the Docker API in a secure way, I propose to add a proxy container (just like is done with Consul), to avoid some issues in Swarm [1].
Furthermore, Mesos Framework authentication should be implemented in Swarm [2]. After some investigation, It appears that adding this will take about 100 lines of Go code.
To make sure that a crash of a single Swarm daemon does not take down all containers started via Swarm, Swarm’s high availability option should be used [3]. I suggest that Swarm should run on each Mesos master node, to provide the same level of availability as Mesos itself.

[1] https://github.com/docker/swarm/issues/1089
[2] https://github.com/docker/swarm/issues/1944
[3] https://docs.docker.com/swarm/multi-manager-setup/

### PR DESCRIPTION
**Pull request by siddharthist**
- [x] Installs cleanly on a fresh build of most recent master branch
- [x] Upgrades cleanly from the most recent release
- [x] Updates documentation relevant to the changes

This is actually branched from #1463 (feature/certificates-role), so it should only be merged after that PR. It's just ready so I thought I'd put it out there. Here's the diff just from this feature: https://github.com/CiscoCloud/mantl/pull/1464/commits/cabb3ecf979965d4c1c36d05cda012ada7064dd1

Acceptance tests (assuming TLS, but should be tested without as well):
From one of the cluster members:
```bash
# docker run --rm --volume=/etc/pki/CA/ca.cert:/etc/pki/CA/ca.cert:ro --volume=/etc/pki/mantl/cert:/etc/pki/mantl/cert:ro --volume=/etc/pki/mantl/key:/etc/pki/mantl/key:ro swarm list --discovery-opt kv.cacertfile=/etc/pki/CA/ca.cert --discovery-opt kv.certfile=/etc/pki/mantl/cert --discovery-opt kv.keyfile=/etc/pki/mantl/key consul://consul.service.consul:8500/nodes
# docker -H :4000 --tlsverify --tlscacert=/etc/pki/CA/ca.cert --tlscert=/etc/pki/mantl/cert --tlskey=/etc/pki/mantl/key run --name=hello-world hello-world
# docker -H :4000 --tlsverify --tlscacert=/etc/pki/CA/ca.cert --tlscert=/etc/pki/mantl/cert --tlskey=/etc/pki/mantl/key logs hello-world
```
From the provisioning machine, ensure that there is only one leader, with the rest as "replicas".
```bash
$ ansible --become role=control -a 'sudo docker -H :4000 --tlsverify --tlscacert=/etc/pki/CA/ca.cert --tlscert=/etc/pki/mantl/cert --tlskey=/etc/pki/mantl/key info' | grep Role
```

Fixes #1406

### CODE REVIEW

File: addons/docker-swarm.yml [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# This playbook has to be copied to the project root directory.
+# See roles/swarm/README.rst for more information.
+
+# CHECK SECURITY - when customizing you should leave this in. If you take it out
+# and forget to specify security.yml, security could be turned off on components
+# in your cluster!
+- include: "{{ playbook_dir }}/playbooks/check-requirements.yml"
```
Comment (sehqlr) [inline addons/docker-swarm.yml]: this should be `playbook_dir }}/../playbooks`

**Comment by siddharthist** [comment]
I cancelled the build because this needs manual testing.

**Comment by siddharthist** [comment]
rebased, still needs manual testing

File: addons/docker-swarm.yml (hunk 1 above)
Comment (siddharthist) [inline addons/docker-swarm.yml]: Update on this: It shouldn't.

This role depends on the certificates role, which references the path `ss/cacert.pem`. This path is relative to the working dir of the playbook, which is why the comment at the top of the file is there.

**Comment by siddharthist** [comment]
I have seen a Consul lock error on this once, but rerunning the playbook fixed it.

**Comment by v6** [comment]
// , Is there anything else you would recommend to keep an eye out for, before trying this out?

**Comment by siddharthist** [comment]
@v6 Nope! Just read the docs :) please let me know how it goes!

**Comment by sehqlr** [comment]
@siddharthist: I was working on testing this, and when I run the docker-swarm playbook, it breaks the UI. I think it might be related to recent reboot problems. Can you confirm?

**Comment by siddharthist** [comment]
@sehqlr https://github.com/CiscoCloud/mantl/pull/1464/files#diff-a622d798ec6462f0d600d2dcafb1a5a2R34

**Comment by sehqlr** [comment]
I completely forgot about that. Once I fixed that, everything worked.

**Comment by siddharthist** [comment]
Thanks @sehqlr!

**Pull request merged**